src/coreclr/tools/aot/ILCompiler.LLVM/Compiler/LLVMCodegenCompilation.cs

```diff
@@ -284,5 +284,10 @@ public override MethodDesc GetCompilerHelpersMethodDesc(string className, string
             MetadataType helperType = TypeSystemContext.SystemModule.GetKnownType("Internal.Runtime.CompilerHelpers", className);
             return helperType.GetKnownMethod(helperMethodName, null);
         }
+
+        public override bool CanTailCall(MethodDesc caller)
+        {
+            return !NeedsReturnStackSlot(caller.Signature);
```

Comment:
The codegen arch specific tailcall limitations are handled in the JIT importer and morph. Can this check be moved there?

Reply:
Thanks,
I've moved some stuff to morph, that looked straightforward enough, I made a few methods in Llvm static and had the usual `min`/`std::min` problem in morph.  What is it that I should be changing in importer.  I had a look at `impTailCallRetTypeCompatible` and probably `allowWidening` should always be false for LLVM.  I'm not sure what else will fail with the compatible check at the moment.  As there are no unsigned types, probably the size check is going to work out the same as an actual LLVM type check.

Reply:
I think morph is a reasonable place for this check. It does similar ABI-dependent checks for other targets as well.